docs/dev/design/secure-api-access-from-builders.rst

```diff
@@ -0,0 +1,130 @@
+Secure API access from builders
+===============================
+
+Goals
+-----
+
+- Provide a secure way for builders to access the API.
+- Limit the access of the tokens to the minimum required.
+
+Non-goals
+---------
+
+- Migrate builds to use API V3
+- Implement this mechanism in API V3
+- Expose it to users
+
+All these changes can be made in the future, if needed.
+
+Current state
+-------------
+
+Currently, we access the API V2 from the builders using the credentials of the "builder" user.
+This user is a superuser, it has access to all projects,
+write access to the API, access to restricted endpoints, and restricted fields.
+
+The credentials are hardcoded in our settings file,
+so if there is a vulnerability that allows users to have access to the settings file,
+the attacker will have access to the credentials of the "builder" user,
+giving them full access to the API and all projects.
+
+Proposed solution
+-----------------
+
+Instead of using the credential of a super user to access the API,
+we will create a temporal token attached to a project, and one of the owners of the project.
+This way this token will have access to the given project only for a limited period of time.
+
+This token will be generated from the webs,
```

Comment:
That sounds alarming :smile: 

But I guess it's the web instance?

Tokens sound fine, and I totally understand their cycle and scope as specified below. But I can't help but wonder if there's something to fetch from the idea of "tickets" as well?

In this case, we could entirely separate this authentication logic from our normal token authentication on API endpoints and design something freely and separately. If it's feasible.

Reply:
> But I guess it's the web instance?

yes, from a web instance, from where we trigger the builds.

> In this case, we could entirely separate this authentication logic from our normal token authentication on API endpoints and design something freely and separately for "tickets". If it's feasible.

We can totally implement our own thing, but if we can re-use an existing package, that will save us a lot of time and complexity, both packages allow us to define our own models, so we can extend them as we want.

Reply:
Cool :+1: Will close this :+1: I asked a couple of elaborated questions in another thread, but I think you know the answers, so I think this looks on track :+1: